File: bindings/python/satox_bindings/core_manager.py
"""
Satox SDK Core Manager Binding
"""

from typing import Dict, List, Optional, Any
import asyncio
from .error import SatoxError
from .types import CoreConfig, CoreStatus

class CoreManager:
    """Wrapper for the C++ CoreManager"""
    
    def __init__(self):
        self._core = None  # Will be set by native binding
        self._initialized = False
        self._callbacks = []
    
    async def initialize(self, config: Optional[CoreConfig] = None) -> bool:
        """Initialize the core manager"""
        if self._initialized:
            raise SatoxError("CoreManager already initialized")
        
        try:
            # Call C++ core implementation
            result = await self._call_native("initialize", config or {})
            self._initialized = result
            return result
        except Exception as e:
            raise SatoxError(f"Failed to initialize CoreManager: {str(e)}")
    
    async def shutdown(self) -> bool:
        """Shutdown the core manager"""
        if not self._initialized:
            return True
        
        try:
            result = await self._call_native("shutdown")
            self._initialized = False
            return result
        except Exception as e:
            raise SatoxError(f"Failed to shutdown CoreManager: {str(e)}")
    
    async def get_status(self) -> CoreStatus:
        """Get core manager status"""
        self._ensure_initialized()
        return await self._call_native("get_status")
    
    def register_callback(self, callback):
        """Register a callback for core events"""
        self._callbacks.append(callback)
    
    def unregister_callback(self, callback):
        """Unregister a callback"""
        if callback in self._callbacks:
            self._callbacks.remove(callback)
    
    def _ensure_initialized(self):
        """Ensure the manager is initialized"""
        if not self._initialized:
            raise SatoxError("CoreManager not initialized")
    
    async def _call_native(self, method: str, *args) -> Any:
        """Call native C++ method"""
        # This will be implemented by the native binding
        raise NotImplementedError("Native binding not implemented") 

File: bindings/typescript/src/core-manager.ts
/**
 * Satox SDK Core Manager Binding
 */

import { SatoxError } from './error';
import { CoreConfig, CoreStatus } from './types';

export class CoreManager {
    private core: any; // Native binding
    private initialized: boolean = false;
    private callbacks: Array<(event: string, success: boolean) => void> = [];

    constructor() {
        // Initialize native binding
        this.core = require('./native/satox_core');
    }

    async initialize(config?: CoreConfig): Promise<boolean> {
        if (this.initialized) {
            throw new SatoxError('CoreManager already initialized');
        }

        try {
            const result = await this.core.initialize(config || {});
            this.initialized = result;
            return result;
        } catch (error) {
            throw new SatoxError(`Failed to initialize CoreManager: ${error.message}`);
        }
    }

    async shutdown(): Promise<boolean> {
        if (!this.initialized) {
            return true;
        }

        try {
            const result = await this.core.shutdown();
            this.initialized = false;
            return result;
        } catch (error) {
            throw new SatoxError(`Failed to shutdown CoreManager: ${error.message}`);
        }
    }

    async getStatus(): Promise<CoreStatus> {
        this.ensureInitialized();
        return await this.core.getStatus();
    }

    registerCallback(callback: (event: string, success: boolean) => void): void {
        this.callbacks.push(callback);
    }

    unregisterCallback(callback: (event: string, success: boolean) => void): void {
        const index = this.callbacks.indexOf(callback);
        if (index > -1) {
            this.callbacks.splice(index, 1);
        }
    }

    private ensureInitialized(): void {
        if (!this.initialized) {
            throw new SatoxError('CoreManager not initialized');
        }
    }
} 